### PR TITLE
Obtain plugin settings in a static context

### PR DESCRIPTION
**Pull request by s1monw**
Today we load the plugins with a fake settings instance and then
fetch all settings level extensions from the plugings. This causes several
consistency issues with plugins sicne they might be instantiated with a
stale settings instance. This also causes issues if plugins need an instance of
`Environment` at construction time which isn't available today but can be made
available in a followup change.

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/plugins/Plugin.java [hunk 1]
```diff
@@ -324,4 +336,38 @@ public final void onModule(ClusterModule module) {}
      */
     @Deprecated
     public final void onModule(DiscoveryModule module) {}
+
+    /**
+     * PluginSettings allow plugins to extend the core by decaring custom settings, adding settings in addition to the defined
```
Comment (rjernst) [inline core/src/main/java/org/elasticsearch/plugins/Plugin.java]: typo: decaring -> declaring

**Review by rjernst — COMMENTED** [review]
I wonder if instead of having a factory class, the PluginsService could have a `load` method which must be called to load the plugins, so this doesn't happen at construction time (then have a SetOnce internally for the loaded Plugin classes).

Looks good otherwise.

**Comment by s1monw** [comment]
@rjernst I pushed new commits. I moved everything into the constructor which was possible after getting rid of the useless baseclass

**Review by rjernst — APPROVED** [review]
LGTM

**Comment by javanna** [comment]
@s1monw is this still relevant?

**Comment by s1monw** [comment]
I am closing this for now. The realm settings are still obtained through SecurityExtensions which blocks this.

**Pull request closed without merging**